#824 Target Bean Factory

remmeier opened the pull request:
Introduced @ObjectFactory annotation to distinguish between factory and mapping methods. So far this was decided based on the lack of parameters. @ObjectFactory lets a factory method carry the same
signature as a mapping method, letting it have access to the source objects.

remmeier commented:
not yet ready to be pulled in. Have a look the changes whether the approach is ok.

I'm not quite sure about the best approach for MappingResolverImpl where I make use setParameters(...). Maybe there is a better way to do the generation of the factorymethod call with freemarker.

Sometimes I get java.lang.NoClassDefFoundError: org/mapstruct/ObjectFactory where I'm not quite sure where this is coming from.

agudian commented on processor/src/main/java/org/mapstruct/ap/internal/model/source/SourceMethod.java:
```diff
@@ -238,7 +241,11 @@ private SourceMethod(Type declaringMapper, ExecutableElement executable, List<Pa
 
-    private Parameter determineMappingTargetParameter(Iterable<Parameter> parameters) {
+    private boolean determineIfIsObjectFactory(ExecutableElement executable) {
+		return executable.getAnnotation(ObjectFactory.class) != null || getTargetTypeParameter() != null || parameters.isEmpty();
```
Use the `ObjectFactoryPrism` here to determine if the executable is annotated with `ObjectFactory`.

Rule of thumb: we don't use any classes from the annotation modules (`mapstruct` and `mapstruct-jdk8`) in the processor (with the exception of the reference in the generator class for the prisms). That way the processor can run without having the annotations on its own classpath. And it usually doesn't have them, which would explain your `NoClassDefFound` errors.

agudian commented on processor/src/main/java/org/mapstruct/ap/internal/model/HelperMethod.java:
```diff
@@ -121,6 +121,11 @@ public Parameter getMappingTargetParameter() {
+    
+    @Override
+   	public boolean isObjectFactory() {
+   		return getTargetTypeParameter() != null || getParameters().isEmpty();
```
just return `false` here.

agudian commented on processor/src/main/java/org/mapstruct/ap/internal/model/source/ForgedMethod.java:
```diff
@@ -216,4 +216,9 @@ public MapperConfiguration getMapperConfiguration() {
+
+    @Override
+   	public boolean isObjectFactory() {
+   		return getTargetTypeParameter() != null || getParameters().isEmpty();
```
just return `false` here. We wouldn't create a forged factory method, at least not yet.

agudian commented on processor/src/main/java/org/mapstruct/ap/internal/model/source/builtin/BuiltInMethod.java:
```diff
@@ -141,6 +141,11 @@ public Parameter getTargetTypeParameter() {
 
+    @Override
+	public boolean isObjectFactory() {
+		return getTargetTypeParameter() != null || getParameters().isEmpty();
```
just return `false` here. We have no built-in methods that act as factory methods.

remmeier commented:
thx for the input, I updated the PR accordingly. The no NoClassDefFoundError have gone away.

I added two further test cases with multiple input sources. I suspect maybe more can be done in that area.

agudian commented on processor/src/test/java/org/mapstruct/ap/test/factories/Bar6Factory.java:
```diff
@@ -0,0 +1,33 @@
+public class Bar6Factory {
+
+    @ObjectFactory
+    public Bar6 createBar6(Foo6A foo6A) {
```
This wouldn't work if you would require only `Foo6B` as argument type, right?

I think somewhere around the Lifecycle methods, we have some code that finds all possible (assignable) combinations for passing source/target arguments to all candidate methods. I think we need to do something very similar here (maybe even reuse the code?), but raise an error in case of multiple matches - at least I can't say from the top of my head how we could select a "best matching" one in case of ambiguities. Well, perhaps favor a factory method with arguments above one without any arguments. But other than that, how would we choose between one that takes Foo6A and one that takes Foo6B?

agudian commented on processor/src/main/java/org/mapstruct/ap/internal/model/MappingMethod.java:
```diff
@@ -86,6 +86,10 @@ protected MappingMethod(Method method, List<LifecycleCallbackMethodReference> be
     }
 
+    public void setParameters(List<Parameter> parameters) {
+        this.parameters = parameters;
```
Please leave the class to be immutable and either add a dedicated construtor or convert the class to use a builder instead.

agudian commented on core-common/src/main/java/org/mapstruct/ObjectFactory.java:
```diff
@@ -0,0 +1,46 @@
+ * EntityManager to check whether the entity already exists and then return its
+ * managed instance.
+ *
+ * @author Remo Meier
```
Could you also add actual code examples here? Over the time we want to convert our javadoc to be more example driven, and I think we already have some in the existing annotations.

agudian replied:
The doc should also mention that the annotation is optional for factory methods without parameters.

agudian commented on core-common/src/main/java/org/mapstruct/ObjectFactory.java:
```diff
@@ -0,0 +1,46 @@
+ * This method allows to mark a method as a factory method to create beans.
+ * The type of the factory method is determined by its return type. A factory is used
+ * to create a bean if their types match. If multiple factories match,
+ * the closest one is chosen.
```
Whatever we end up with, we should explain what "closest one" is referring to 🙂.

agudian commented:
Dear Jenkins, ok to test!

remmeier commented:
done, 

Parameter assigment is now shared with the lifecycle methods. If there are multiple factory methods, it chosen the one with a valid assigment. If multiple ones would be possible, an exception is thrown. The only exception is the factory method without parameters, that one is only chosen if no other factory is found.

otherwise:
- some smaller doc udpates.
- more test cases

agudian commented:
Nice, I've just checked it out and played around with the tests a bit. Feels really good.

Two more things we should support, I think: a `@TargetType` argument would be cool:
```
   @ObjectFactory
    public <T> T createGeneric(Foo7B foo7B, @TargetType Class<T> clazz) {
        return ...;
    }
```
That's a pretty easy part. In `MappingResolverImpl.getFactoryMethod`, you can add a `Parameter` for the target type and pass that together with the list of the source parameters into `getParameterAssignments`:
```
        List<Parameter> availableParams = new ArrayList<Parameter>( mappingMethod.getSourceParameters() );
        availableParams.add( new Parameter(
            null,
            typeFactory.classTypeOf( targetType ),
            false,
            true,
            false ) );
```

Then there's another thing that we should have to help users in case they face the *ambigious factory method* error message, and that's supporting a qualifier annotation (via `@BeanMapping(qualifiedBy = ..)` or `@BeanMapping(qualifiedByName = ..)`). Or does that work already? In any case, the qualifier would also have to overrule the factory-method-with-params over factory-method-without-params logic... And I didn't try that part out either, so it might work already 😀 .

Then we also need to add some tests below the package `org.mapstruct.ap.test.erroneous` that check the various error scenarios and the proper generation of the error messages (I've seen them while playing around, but we should have some tests there as well).

It would be great if you could follow up some more with this to bring it to the finish line, if you have the time. We'd really appreciate it.

gunnarmorling commented:
>  in case they face the ambigious factory method error message, and that's supporting a qualifier annotation (via @BeanMapping(qualifiedBy = ..) or @BeanMapping(qualifiedByName = ..))

+1 I'd prefer explicit resolution using a qualifier over the suggested logic (it seems not intuitive that one should be preferred over the other).

remmeier commented:
I committed something to clarify the qualifier (not usable currently).

There is a Bar10Factory and SourceTargetMapperAndBar2Factory has different methods mapping to Bar10. Is it intended like this? I did not yet really figure out how all those qualifiers are processed (and did also not use those in the past). MappingResolveRimpl.getFactoryMethod(...) does get called with SelectionParameters == null for foo10ToBar10Upper. I expected to see Bar10Qualifier. Looking at the code, it seems like I would have to use the BeanMapping annotation, which is then passed to MappingOptions and finally to the getFactoryMethod(...)  to get something different? Any hints how I have to proceed?

Once that is done, the the new qualifier attributes in ObjectFactory need to be considered for the selection process.

agudian commented:
Well, you've found us a documentation gap: `@BeanMapping#qualifiedBy` and `@BeanMapping#qualifiedByName` are not described with regards to factory methods in our reference docs.
But the Javadoc on those attributes gives a hint and we have an example in the tests (`org.mapstruct.ap.test.selection.qualifier.MovieFactoryMapper`). For your test, this would mean that you would the define the mapping methods the following way:

```java
    @BeanMapping(qualifiedBy = Bar10Qualifier.class)
    public abstract Bar10 foo10ToBar10Upper(Foo10 foo10);
    
    @BeanMapping(qualifiedByName = "Bar10NamedQualifier")
    public abstract Bar10 foo10ToBar10Camel(Foo10 foo10);
```
and the factory methods:
```java
    @Bar10Qualifier
    @ObjectFactory
    public Bar10 createBar10Upper(Foo10 foo10) {
        return new Bar10( foo10.getProp().toUpperCase() );
    }

    @Named("Bar10NamedQualifier")
    @ObjectFactory
    public Bar10 createBar10Camel(Foo10 foo10) {
        return new Bar10( foo10.getProp().charAt( 0 ) + foo10.getProp().toLowerCase().substring( 1 ) );
    }
```

So `@ObjectFactory` wouldn't have the attributes `qualifiedBy` and `qualifiedByName`. That's the mechanism used for mapping methods and the simple factory methods and it should apply here just the same.
As for the selection mechanism, the information about the settings of the qualifier attributes in `@BeanMapping` should be part of the `SelectionParameters` and evaluated in the `QualifierSelector`. So this might work already...?

As a side note for us, we should really refactor that `FactoryTest` after we're done with this PR... it's way too big for some time now and I bet the size of it already makes debugging your cases quite time consuming... 😉

remmeier commented:
I see, the other way around with the annotation. Yes, in this case everything works already. I added nevertheless the testcases, added support for TargetType and a ambiguous test.

agudian commented:
Okay, I have just two more tiny comments. Would be great if you could rebase to the head of the master branch and squash your commits into one.
Then this should be ready to merge from my side. 👍

sjaakd commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -1486,6 +1487,28 @@ public class CarMapperImpl implements CarMapper {
+Annotating a factory method further with `@ObjectFactory` lets you gain access to the mapping sources. 
+Source objects can be added as parameters in the same way as for mapping method. The `@ObjectFactory`
+annotation is necessary to let MapStruct know that the given method is only a factory method, not a 
+complete mapping method. 
```
mapping method..  (leave complete out)

sjaakd commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -1486,6 +1487,28 @@ public class CarMapperImpl implements CarMapper {
 ----
 ====
 
+Annotating a factory method further with `@ObjectFactory` lets you gain access to the mapping sources. 
```
further == additionally?

sjaakd replied:
Or better: In addition, annotating a factory method with..

sjaakd commented on processor/src/test/java/org/mapstruct/ap/test/erroneous/ambiguousannotatedfactorymethod/AmbiguousAnnotatedFactoryTest.java:
```diff
@@ -0,0 +1,59 @@
+import org.mapstruct.ap.testutil.runner.AnnotationProcessorTestRunner;
+
+/**
+ * @author Sjaak Derksen
```
I kind of like the current one :smile:

sjaakd commented:
> could you guys please check out the API change and voice any concerns, please?

I'm ok 👍 ..

sjaakd commented:
> could you guys please check out the API change and voice any concerns, please?

Perhaps one general remark.. `ObjectFactory` sounds like an annotation that has more capabilities than creating target objects from source parameters. Should we be more precise in the name or do you see a broader use later on? To name one: I experimented a while ago with having a factory for mappers themselves.

gunnarmorling commented:
Are these factory methods supported for other component models (CDI, Spring), too? I.e. could a factory type providing a parameterized factory method be obtained via DI and `Mapper#uses()`?

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/model/ParameterAssignmentUtil.java:
```diff
@@ -0,0 +1,93 @@
+        return null;
+    }
+
+    private static List<Parameter> findCandidateParameters(List<Parameter> candiateParameters, Parameter parameter) {
```
candiateParameters=candidateParameters

What is a candidate parameter? Some description would be nice.

remmeier commented:
done:
- split of test cases
- minor doc updates as suggested
- added @since 1.2
- simplified factory selection (no constructor factory handled like all the others)

squash to follow

remmeier commented:
I did not touch any of the CDI/Spring related things. I suppose if it work before, it works know.

agudian commented:
💯 🎉 
Merged. Thanks a lot, Remo! It's a great addition! Chime in again anytime if you find something else missing!

I've also pushed a commit to add your name to the list of contributors.

gunnarmorling commented:
Very cool, thanks @remmeier et al.!

sjaakd commented:
My congratulations as well. Nice piece of work!

remmeier commented:
thx as well. 

Is there any roadmap for a first 1.2 CR release? but there is no hurry, I can always release it locally.

gunnarmorling commented:
Hi Remo,

we haven't any fixed schedule for the 1.2 release yet (it will definitely
be much quicker than the 1.1 release). For the time being you can either
build yourself locally or download the snapshot from the Sonatype repo:


https://oss.sonatype.org/content/repositories/snapshots/org/mapstruct/mapstruct-processor/1.2.0-SNAPSHOT/
